### PR TITLE
Add protocolbuffers/gofeatures module

### PR DESCRIPTION
**Pull request by jhump**

### CODE REVIEW

File: modules/static/protocolbuffers/gofeatures/buf.yaml
Comment (pkwarren) [inline modules/static/protocolbuffers/gofeatures/buf.yaml:2]: Do we want to name this `protobuf-go` instead of `gofeatures` in case they ever add more go-specific proto files in the future?

It could be confusing that wellknowntypes contains java/cpp features but the go ones are defined elsewhere.

File: modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf
Comment (pkwarren) [inline modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: Should this have a dependency on protocolbuffers/wellknowntypes? If so does it need to pin a version of the WKTs in the buf.yaml file?

File: modules/sync/protocolbuffers/gofeatures/cas/8fd51058773ac1c2ee378a04f3776fb374ae29813af9b41eac55373543f62effe422b29e76fdd41c0f540fdebe632794cc5b69f2b7a4226b9a1ce63f563a339f
Comment (pkwarren) [inline modules/sync/protocolbuffers/gofeatures/cas/8fd51058773ac1c2ee378a04f3776fb374ae29813af9b41eac55373543f62effe422b29e76fdd41c0f540fdebe632794cc5b69f2b7a4226b9a1ce63f563a339f:9]: It looks like we need to regenerate these files - this doesn't match the buf.md above as it includes content from `cel-spec`.

File: modules/static/protocolbuffers/gofeatures/buf.yaml
Comment (jhump) [reply modules/static/protocolbuffers/gofeatures/buf.yaml:2]: > Do we want to name this `protobuf-go` instead of `gofeatures`.

🤷. I don't anticipate anything else being added here. This is a core Protobuf runtime, so I wouldn't expect it to need any definitions other than custom options and features related to Go code gen, in which case "gofeatures" seems fine. The upside to this name is that it is more descriptive as to what it actually contains. The downside, as you mentioned, is if they eventually add other stuff to it, suddenly making `gofeatures` a bit of a misnomer.

> It could be confusing that wellknowntypes contains java/cpp features but the go ones are defined elsewhere.

Yep. But that's the state of the world with the authoritative sources, too: the core protoc distribution does _not_ include the `go_features.proto` file. The Protobuf team said it was intentional: since `protoc-gen-go` is a separate/independent plugin from the core languages implemented by `protoc`, they are using the approach that any other plugin author would need to take to provide feature definitions. This is much like any other plugin, where the sources for custom options, etc, must be distributed in parallel to the plugin itself.

Also, this file is revisioned with the protobuf-go repo, not with the main protoc releases, so trying to bundle it inside the wellknowntypes module would be problematic from that perspective (not to mention the fact that the current fetch mechanism is not flexible enough to accommodate a single module whose contents are pulled from multiple repos).

FWIW, the Protobuf team only intend to include features for languages implemented by `protoc` in the core release (and thus mirrored into our `protocolbuffers/wellknowntypes` module).

File: modules/sync/protocolbuffers/gofeatures/cas/8fd51058773ac1c2ee378a04f3776fb374ae29813af9b41eac55373543f62effe422b29e76fdd41c0f540fdebe632794cc5b69f2b7a4226b9a1ce63f563a339f
Comment (jhump) [reply modules/sync/protocolbuffers/gofeatures/cas/8fd51058773ac1c2ee378a04f3776fb374ae29813af9b41eac55373543f62effe422b29e76fdd41c0f540fdebe632794cc5b69f2b7a4226b9a1ce63f563a339f:9]: Doh, my bad. I fixed a copy+pasta bug in the sources and forgot to re-generate 🤦

File: modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf
Comment (pkwarren) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: I think this should depend on WKT - without it I got the following error running the fetch script locally (with buf v1.31.0):
```
google/protobuf/go_features.proto:17:28:extension pb.go: tag 1002 is not in valid range for extended type google.protobuf.FeatureSet
```

Comment (jhump) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: Hmm, unclear. The wellknowntypes never require an explicit dep since `buf` bundles them. They are only needed as a dependency to explicitly pin a particular version.

In this case, this version (1.34.0) requires v25.0 or higher of the wellknowntypes. And the _next_ version will require v27.0 or higher. But will that be weird to pin it? Would that mean that all modules that _use_ the gofeatures end up getting "stuck" on whatever version it pins? (I.e. they wouldn't end up using newer wellknowntypes, even if using a newer version of the CLI that bundles a newer version of them.)

File: modules/static/protocolbuffers/gofeatures/buf.yaml
Comment (bufdev) [reply modules/static/protocolbuffers/gofeatures/buf.yaml:2]: I think on balance, I think we keep it as `gofeatures`. This gets back to us choosing to have conflicting namespaces between plugins and modules, I choice I regret. If we wanted to name this consistently, we'd actually just name it `protocolbuffers/go`, but that's not possible since that is taken by the plugin. What we should really do is come up with a naming scheme for repos that share (or could share) plugins and modules, which will be common in the case of options.

With all that typed out...would be open to such a scheme. For example, `buf.build/foo/bar` and `buf.build/foo/bar-api` is one option, but I don't like that `-api` is redundant.

File: modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf
Comment (bufdev) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: I agree with Philip here, let's make it a dependency

Comment (jhump) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: What's our stance on the version to pin: should it be the _minimum_ version required (v25.0)? Or should we keep it recent (e.g. v26.1)?

I'm going to guess the minimum one since it doesn't look like we have anything automated to update it as newer versions of deps are released.

Comment (pkwarren) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: I think we can leave it unpinned if we want - it will just point at the latest WKT version when we sync on the user's system. That also means that we don't have to remember to bump the pinned version in the future.

Comment (jhump) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf:2]: > it will just point at the latest WKT version when we sync on the user's system.

But if the user has already pinned the WKT module, but an older version, it will fail to build, right? I'm also wondering if we should do something different than leaving it unpinned in the next version (which will require a new v27+ of the wellknowntypes).

In fact, that makes me wonder if maybe we should hold off on trying to merge this until we have a v27 of wellknowntypes and a v1.35 or v1.34.1 (whatever they number it) for protobuf-go and not bother trying to sync the current v1.34.0...

**Review by pkwarren — APPROVED** [review]

File: modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf
Comment (jhump) [reply modules/sync/protocolbuffers/gofeatures/cas/66220fa6f78c2ff242ac897567ed4255f1bd5f76bf5e252edc2d72945e75504e73da2dfe9dce502d6e522a0242d2c3a58482836bfc0463f2a460fea985f99ccf]: Now that v27.0 of protocolbuffers/wellknowntypes is available, I've updated this so that it will start with v1.34.1 (which requires v27 of descriptor.proto). Updated the config in this PR.

I'll now start going through the checklist for new modules.

**Comment by jhump** [comment]
I ran the `fetch` action for this branch and it produced #567. The files there LGTM. @pkwarren, want to take one last look at this and #567 before I merge this?

**Review by pkwarren — APPROVED** [review]

File: modules/static/protocolbuffers/gofeatures/rsync.incl
Comment (bufdev) [inline modules/static/protocolbuffers/gofeatures/rsync.incl:2]: Doesn't this have a directory path associated with it?

Comment (pkwarren) [reply modules/static/protocolbuffers/gofeatures/rsync.incl]: Added full import path and appears to work fine - https://github.com/bufbuild/modules/pull/572.

File: modules/static/protocolbuffers/gofeatures/rsync.incl
Comment (bufdev) [inline modules/static/protocolbuffers/gofeatures/rsync.incl:3]: What are these stars for?

Comment (pkwarren) [reply modules/static/protocolbuffers/gofeatures/rsync.incl:3]: I believe it is to recurse in subdirectories but not positive. I can try removing it to see if it still works.

Comment (pkwarren) [reply modules/static/protocolbuffers/gofeatures/rsync.incl:3]: Removing that causes the fetch script to fail with `Failure: "buf.build/protocolbuffers/gofeatures" had no .proto files`.

Comment (bufdev) [reply modules/static/protocolbuffers/gofeatures/rsync.incl:3]: What does the rsync manual say?

Comment (pkwarren) [reply modules/static/protocolbuffers/gofeatures/rsync.incl:3]:
```

       Note that, when using the --recursive (-r) option (which is implied by -a), every subcomponent of every path is visited from the top down, so include/exclude patterns get applied recursively to each subcomponent's
       full name (e.g. to include "/foo/bar/baz" the subcomponents "/foo" and "/foo/bar" must not be excluded).  The exclude patterns actually short-circuit the directory traversal stage when rsync finds the files to send.
       If a pattern excludes a particular parent directory, it can render a deeper include pattern ineffectual because rsync did not descend through that excluded section of the hierarchy.  This is particularly important
       when using a trailing '*' rule.  For instance, this won't work:

              + /some/path/this-file-will-not-be-found
              + /file-is-included
              - *

       This fails because the parent directory "some" is excluded by the '*' rule, so rsync never visits any of the files in the "some" or "some/path" directories.  One solution is to ask for all directories in the
       hierarchy to be included by using a single rule: "+ */" (put it somewhere before the "- *" rule), and perhaps use the --prune-empty-dirs option.  Another solution is to add specific include rules for all the parent
       dirs that need to be visited.  For instance, this set of rules works fine:

              + /some/
              + /some/path/
              + /some/path/this-file-is-found
              + /file-also-included
              - *
```

Specifically:

> One solution is to ask for all directories in the hierarchy to be included by using a single rule: "+ */" (put it somewhere before the "- *" rule)

Comment (bufdev) [reply modules/static/protocolbuffers/gofeatures/rsync.incl:3]: Can you translate that - is there a tighter way to do this, or why or why not does this guarantee we are just limited to the two files we want?

Comment (pkwarren) [reply modules/static/protocolbuffers/gofeatures/rsync.incl:3]: If we wanted to be the most strict, then we could add these lines:

```
google/
google/protobuf/
```

However, using `*/` is a shortcut to not have to include those, since we're using the `--prune-empty-dirs` rsync option.

**Review by bufdev — APPROVED** [review]

**Pull request merged**
